### PR TITLE
[BackEnd] Refactor Images Server

### LINKED ISSUE #239: Improve book add input parsing/error checking
Specifically for ISBNs, when an ISBn was tried to submit followed by a comma, there is an error message saying it can't be added, when this is a valid input that should properly work. In fact, I believe it does work, but the error message still shows up. This needs to be resolved.

### LINKED ISSUE #492: Refactor Image Server to use Local Image Server Instance

### PR DESCRIPTION
**Pull request by hkder**
## Feature Description/Implementation
Book LookUp now checks for ending commas. Closes #239

Closes #492 
Endpoint Change
- setDefaultImage -> sets the image related to the book with the default image
- image_bytes -> used generally for custom user given image files
- image_url -> input the url of the image you want to give (normally used when books are newly added)

Note) if both are provided precedence is given to image_bytes.
Note) For Ghost books the book lookup will show as if it is a new book. Just use add endpoint and the backend will handle the case.

## Dependencies
/srv/media/books must be present - the devops repo takes care of this

## Everything Else
The code will be refactored as Book Add and Book Modify share lots of duplicated code.
This PR is for fast frontend integration.

### CODE REVIEW

File: backend/books/views.py [hunk 1]
```diff
@@ -174,21 +183,43 @@ def create(self, request, *args, **kwargs):
...
         if setDefaultImage:
             url = self.isbn_toolbox.get_default_image_url()
-        else:
+        elif self.has_image_bytes(request):
             url = self.isbn_toolbox.commit_image_raw_bytes(request, book[0].id, isbn_13)
+        elif self.has_image_url(request):
+            url = self.isbn_toolbox.commit_image_url(request, book[0].id, isbn_13)
 
         obj, created = BookImage.objects.get_or_create(
             book_id=book[0].id,
```
Comment (danielfeinblatt) [inline backend/books/views.py:225]: If none of these conditions are met, then `url` would be undefined. So either, why will one of these conditions always guaranteed to be met, or need to add some type of `else`

File: backend/books/views.py [hunk 2]
```diff
@@ -361,14 +396,27 @@ def convert_zero_to_null(self, data):
...
         book = Book.objects.filter(isbn_13=isbn_13)
 
         # This creates an image in static and sends a file
         if setDefaultImage:
             url = self.isbn_toolbox.get_default_image_url()
-        else:
+        elif self.has_image_bytes(request):
             url = self.isbn_toolbox.commit_image_raw_bytes(request, book[0].id, isbn_13)
+        elif self.has_image_url(request):
+            url = self.isbn_toolbox.commit_image_url(request, book[0].id, isbn_13)
```
Comment (danielfeinblatt) [inline backend/books/views.py:419]: Same as above with meeting these conditions otherwise `url` is undefined

**Review by danielfeinblatt — CHANGES_REQUESTED** [review]
Left some comments and questions

File: backend/books/isbn.py [hunk 3]
```diff
@@ -202,28 +210,42 @@ def create_local_image(self, filename_without_extension, image_bytes):
...
             absolute_location = ''
 
         return absolute_location, filename # Absolute location is used to send the static file to image server
+
+    def commit_image_url(self, request, book_id, isbn_13):
+        # Get the image url
+        end_url = request.data.get('image_url')
+
+        # If the url is the default iamge url no need to download
+        if end_url == self.get_default_image_url():
```
[the remark this answers is not in the document]
Comment (hkder) [reply backend/books/isbn.py:227]: get_image_raw_bytes takes care if the given URL is invalid. Main purpose here is when the user clicks remove image and I do not want to redownload the default image already in the image server.

File: backend/books/views.py [hunk 4]
```diff
@@ -35,23 +35,20 @@ class ISBNSearchView(APIView):
...
         serializer = ISBNSerializer(data=request.data)
         serializer.is_valid(raise_exception=True)
         uri = request.build_absolute_uri()
 
-        self.preprocess(uri)
+        # Set the internal image base url using the request
+        self.isbn_toolbox.set_internal_image_base_url(uri)
 
         # Split ISBN with spaces and/or commas
         raw_isbn_list = re.split("\s?[, ]\s?", serializer.data['isbns'].strip())
```
[the remark this answers is not in the document]
Comment (hkder) [reply backend/books/views.py:47]: addressed in issue #511

File: backend/books/views.py (hunk 1 above)
Comment (hkder) [reply backend/books/views.py:225]: line 186 `if self.handle_image(request, setDefaultImage):`

File: backend/books/views.py (hunk 2 above)
Comment (hkder) [reply backend/books/views.py:419]: answered above

File: backend/books/isbn.py [hunk 5]
```diff
@@ -202,28 +210,42 @@ def create_local_image(self, filename_without_extension, image_bytes):
         try:
             image = Image.open(io.BytesIO(image_bytes))
             filename = f'{filename_without_extension}.{image.format.lower()}'
-            absolute_location = f'{settings.STATICFILES_DIRS[0]}/{filename}'
+            absolute_location = f'{self._internal_book_image_absolute_path}/{filename}'
             image.save(absolute_location)
-        except:
+        except Exception as e:
+            print(e)
```
[the remark this answers is not in the document]
Comment (hkder) [reply backend/books/isbn.py]: removed

File: backend/books/utils.py [hunk 6]
```diff
@@ -53,11 +51,11 @@ def get_port_number(url):
     return url.split(":")[2].split("/")[0]
 
 def url_to_static_image_service(url):
-    return '/'.join(url.split('/')[:5])
+    return '/'.join(url.split('/')[:3])
```
[the remark this answers is not in the document]
Comment (hkder) [reply backend/books/utils.py:57]: added

**Review by danielfeinblatt — APPROVED** [review]
LGTM

**Pull request merged**